ext/statsd/statsd.c

```diff
@@ -0,0 +1,376 @@
+#include <stdbool.h>
+#include <ruby.h>
+#include <ruby/encoding.h>
+#include <ruby/st.h>
+
+#define DATAGRAM_SIZE_MAX 4096
+#define SAMPLE_RATE_SIZE_MAX 16
+#define NORMALIZED_TAGS_CACHE_ENABLED 1
+#define NORMALIZED_TAGS_CACHE_MAX 512
+#define NORMALIZED_NAMES_CACHE_ENABLED 1
+#define NORMALIZED_NAMES_CACHE_MAX 512
+
+static ID idTr, idNormalizeTags, idDefaultTags, idPrefix;
+
+struct datagram_builder {
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_table *normalized_tags_cache;
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  st_table *normalized_names_cache;
+#endif
+  VALUE str_normalize_chars;
+  VALUE str_normalize_replacement;
+  // cached default tags ivar to skip a lookup
+  VALUE default_tags;
+  bool empty_default_tags;
+  int prefix_len;
+  int len;
+  // last member to not glob up cache lines to access other struct members
+  char datagram[DATAGRAM_SIZE_MAX];
+};
+
+// GC callback to mark the wrapper struct. Conditionally symbol tables if caching is enabled (values only)
+// and the cached default tags as well.
+void
+datagram_builder_mark(void *ptr)
+{
+  const struct datagram_builder *builder = (struct datagram_builder *)ptr;
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  rb_mark_tbl(builder->normalized_tags_cache);
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  rb_mark_tbl(builder->normalized_names_cache);
+#endif
+  rb_gc_mark(builder->str_normalize_chars);
+  rb_gc_mark(builder->str_normalize_replacement);
+  rb_gc_mark(builder->default_tags);
+}
+
+// GC callback to free the wrapper struct. Conditionally symbol tables if caching is enabled
+// and the struct itself.
+void
+datagram_builder_free(void *ptr)
+{
+  struct datagram_builder *builder = (struct datagram_builder *)ptr;
+  if (!builder) return;
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_free_table(builder->normalized_tags_cache);
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  st_free_table(builder->normalized_names_cache);
+#endif
+  xfree(builder);
+}
+
+// Be nice to ObjectSpace and let the size be known. We'd likely want some feedback on
+// this with various normalized cache size values.
+size_t
+datagram_builder_size(const void *ptr)
+{
+  size_t size;
+  const struct datagram_builder *builder = (struct datagram_builder *)ptr;
+  size = sizeof(struct datagram_builder);
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  size += st_memsize(builder->normalized_tags_cache);
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  size += st_memsize(builder->normalized_names_cache);
+#endif
+  return size;
```

Comment:
Will this include the sizes of the `VALUE` members on the builder (assuming the values can’t be embedded)?

Reply:
My understanding of object memory size reporting (as expected to be returned by `ObjectSpace.memsize_of`) is that retained memory shouldn't be included. See [Class](https://github.com/ruby/ruby/blob/12009fb3b984faf346250ee7723ecccf56622ce0/gc.c#L3937-L3954) for example 

Heap dumps follow that model too, otherwise the [dominator tree](https://github.com/csfrancis/harb/pull/6) for harb would not have been needed :smile:

[st_memsize](https://github.com/ruby/ruby/blob/0f2f07a9bb0fcbfe897e956965fcf0cfb1ffdbfa/st.c#L647-L654) is just memory consumed by the hash table internals (bins and entries)